File: zmy/cpp/3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        std::unordered_map<char, int> mmap;
        int start = 0;
        int longest = 0;
        for (int i=0; i<s.length(); i++){
            if(mmap.find(s[i])!=mmap.end() && mmap[s[i]] >= start) {
                start = mmap[s[i]] + 1; 
            }
            if(i-start +1 > longest) {
                longest = i - start +1;
            }
            mmap[s[i]] = i;
        }
        return longest;
    }
};

File: zmy/golang/3.longest-substring-without-repeating-characters.go
/*
 * @lc app=leetcode id=3 lang=golang
 *
 * [3] Longest Substring Without Repeating Characters
 */
func lengthOfLongestSubstring(s string) int {
	mmap := make(map[rune]int)
	start := 0
	longest := 0
	for i, ch := range s {
		if last1, ok := mmap[ch]; ok && last1 >= start {
			start = last1 + 1
		}
		if i-start + 1 > longest {
			longest = i-start + 1
		}
		mmap[ch] = i
	}
	return longest
}

